lua/neogit/diff.lua

```diff
@@ -0,0 +1,166 @@
+local M = {}
+local a = require("neogit.async")
+
+local state = {
+  open = false,
+  dec = nil,
+  lhs = nil,
+  rhs = nil
+}
+
+M.state = state
+
+function M.invoke_mapping(name, mode, key)
+  M.mappings[name][mode .. " " .. key]()
+end
+
+function M.close()
+  if state.dec ~= nil then
+    vim.api.nvim_buf_call(state.dec.buf, function()
+      vim.cmd [[au! * <buffer>]]
+    end)
+    vim.api.nvim_win_close(state.dec.win, false)
+    state.dec = nil
+  end
+  if state.lhs ~= nil then
+    vim.api.nvim_buf_call(state.lhs.buf, function()
+      vim.cmd [[au! * <buffer>]]
+    end)
+    vim.api.nvim_win_close(state.lhs.win, false)
+    state.lhs = nil
+  end
+  if state.rhs ~= nil then
+    vim.api.nvim_buf_call(state.rhs.buf, function()
+      vim.cmd [[au! * <buffer>]]
+    end)
+    vim.api.nvim_win_close(state.rhs.win, false)
+    state.rhs = nil
+  end
+  state.open = false
+end
+
+M.open = a.sync(function(lhs_info, rhs_info)
```

Comment:
Is this method async for future use? There are no asynchronous calls in there except for textlock schedules, so you could also just not make this function async and wait for the textlock before calling it. Might also fix the oddity at the end of the function.

Reply:
Yeah I only made the function async to be able to use wait_for_textlock. I guess you are right that the responsiblity should lie on the caller side.